pandas/core/generic.py

```diff
@@ -1948,6 +1948,119 @@ def tail(self, n=5):
             return self
         return self.iloc[-n:]
 
+  
+    def sample(self, n=None, frac=None, replace=False, weights=None, random_state=None, axis = 0):
+        """
+        Returns a random sample of rows from object. 
+        
+        Parameters
+        ----------
+        n : int, optional
+            Number of rows to return. Cannot be used with `frac`.
+            Default = 1 if `frac` = None. 
+        frac : float, optional
+            Share of rows to return. Cannot be used with `n`. 
+        replace : boolean, optional
+            Sample with or without replacement. Default = False. 
+        weights : str or ndarray-like, optional
+            Default 'None' results in equal probability weighting. 
+            If called on a DataFrame or Panel, will also accept the name of a 
+            column as a string. Must be same length as index. 
+            If weights do not sum to 1, they will be normalized to sum to 1. 
+            Missing values in the weights column will be treated as zero. 
+            inf and -inf values not allowed. 
+        random_state : int or numpy.random.RandomState, optional
+            Seed for the random number generator (if int), or numpy RandomState 
+            object.
+        axis : int or string, optional
+            Axis to sample. Accepts axis number or name. Default = 0.
+            
+        Returns
+        -------
+        Same type as caller. 
+        """
+        
+        ###        
+        # Processing axis argument
+        ###
+        
+        # Check validity of axis argument.
+        axis = self._get_axis_number(axis)
+    
+        # Store length of relevant axis of object. 
+        axis_length = self.shape[axis]    
+    
+        ###
+        # Clean / process random_state argument
+        ###
+     
+        rs = com._random_state(random_state)
+    
+        ###
+        #  Process weight argument
+        ###
+    
+        # Check weights for compliance
+        if weights is not None:
+                                      
+            # Strings acceptable if not a series
+            if isinstance(weights, string_types): 
+
+                if self.ndim > 1 :
+                    try:
+                        weights = self[weights]
```

Comment:
this should access the 'weights' column in the correct axis. Eg if you sample a dataframe it's columns (`axis=1`), should it then be possible to give a row index name? 
In any case, now this can trigger the "Weights and axis to be sampled must be of same length" error when the dataframe is not square (and would give faulty results if it would be square)